Delivery methods without design flaws

litetex opened the pull request:
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [ ] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [x] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API.

Fixes #858 

Status: Work in progress

Stypox commented:
Took a brief look: I like the structure. I usually tend to avoid having this many abstract things, but in this case there is no way to avoid it because of the complexity of the data. With this structure a user of the extractor would need to use `instanceof` to decide how to play a `Stream` he obtained, doesn't he?

Also, what about removing completely the concept of `getVideoStreams`, `getAudioStreams` and `getAudioVideoStreams` and just create a `getStreams` that returns all of them, and then the library user can just call `.filter(VideoStream.class::isInstance).map(VideoStream.class::cast)` to filter out what he needs (we could provide this as a utility method)? This way we would avoid duplicate code for extracting the same thing in 3 places, and we would stop having to store temporarily extracted streaming data in fields that use RAM needlessly. Anyway, this is a suggestion for a future PR.

litetex commented:
@Stypox 
Sound's interesting but as you said I think this is best in a followup.

I'm currently already struggling with NewPipe itself because nearly everything that is player related has to be reworked (which with the fact that I currently don't have much spare time gradually slows down the progress quite a bit).

Btw there will be further unfinished changes inbound here that are mostly download-related e.g. marking TorrentData as not downloadable or how the playback should be resolved in the best way, e.g. preferring the hlsMasterPlaylistUrl or something else for certain streams.

Stypox commented:
@litetex any news? :-)

litetex commented:
@Stypox 
Current state: I'm still working on it (but I'm currently busy doing other stuff so progress is kind of slow). The extractor side should be nearly finished (did a few redesigns regarding downloads) but the client side still needs major changes.
For now I rebased the branch and fixed conflicts from the last few months.
